xbmc/windowing/gbm/GBMUtils.cpp

```diff
@@ -0,0 +1,555 @@
+/*
+ *      Copyright (C) 2005-2013 Team XBMC
+ *      http://xbmc.org
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <drm/drm_mode.h>
+#include <EGL/egl.h>
+#include <unistd.h>
+
+#include "WinSystemGbmGLESContext.h"
+#include "guilib/gui3d.h"
+#include "utils/log.h"
+#include "settings/Settings.h"
+
+#include "GBMUtils.h"
+
+static struct drm *m_drm = new drm;
+static struct drm_fb *m_drm_fb = new drm_fb;
+
+static struct gbm_bo *m_bo = nullptr;
+static struct gbm_bo *m_next_bo = nullptr;
+
+static drmModeResPtr m_drm_resources = nullptr;
+static drmModeConnectorPtr m_drm_connector = nullptr;
+static drmModeEncoderPtr m_drm_encoder = nullptr;
+static drmModeCrtcPtr m_orig_crtc = nullptr;
+
+static struct pollfd m_drm_fds;
+static drmEventContext m_drm_evctx;
+static int flip_happening = 0;
+
+bool CGBMUtils::InitGbm(RESOLUTION_INFO res)
+{
+  GetMode(res);
+
+  m_drm->gbm->width = m_drm->mode->hdisplay;
+  m_drm->gbm->height = m_drm->mode->vdisplay;
+
+  m_drm->gbm->surface = gbm_surface_create(m_drm->gbm->dev,
+                                           m_drm->gbm->width,
+                                           m_drm->gbm->height,
+                                           GBM_FORMAT_ARGB8888,
+                                           GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
+
+  if(!m_drm->gbm->surface)
+  {
+    CLog::Log(LOGERROR, "CGBMUtils::%s - failed to create surface", __FUNCTION__);
+    return false;
+  }
+
+  CLog::Log(LOGDEBUG, "CGBMUtils::%s - created surface with size %dx%d", __FUNCTION__,
+                                                                         m_drm->gbm->width,
+                                                                         m_drm->gbm->height);
+
+  return true;
+}
+
+void CGBMUtils::DestroyGbm()
+{
+  if(m_drm->gbm->surface)
+  {
+    gbm_surface_destroy(m_drm->gbm->surface);
+  }
+
+  m_drm->gbm->surface = nullptr;
+}
+
+bool CGBMUtils::SetVideoMode(RESOLUTION_INFO res)
+{
+  GetMode(res);
+
+  gbm_surface_release_buffer(m_drm->gbm->surface, m_bo);
```

Comment:
This is too early; you only want to release the buffer after drmModeSetCrtc has returned.